# Sistema de detección de enlaces spam

---

Queremos implementar un sistema que sea capaz de detectar automáticamente si una página web contiene spam o no basándonos en su URL

In [17]:
import pandas as pd
import matplotlib.pyplot as plt

---

## Paso 1: Carga del conjunto de datos

In [9]:
df = pd.read_csv("https://raw.githubusercontent.com/4GeeksAcademy/NLP-project-tutorial/main/url_spam.csv")
df.head()

url  is_spam
0  https://briefingday.us8.list-manage.com/unsubs...     True
1                             https://www.hvper.com/     True
2                 https://briefingday.com/m/v4n3i4f3     True
3   https://briefingday.com/n/20200618/m#commentform    False
4                        https://briefingday.com/fan     True

---

## Paso 2: Preprocesa los enlaces

Utiliza lo visto en este módulo para transformar los datos para compatibilizarlos con el modelo que queremos entrenar. Segmenta las URLs en partes según sus signos de puntuación, elimina las stopwords, lematiza, etcétera.

Asegúrate de dividir convenientemente el conjunto de datos en train y test como hemos visto en lecciones anteriores.

### 2.1 Transformación categórica a numérica

Transformamos la categoría `is_spam` en valores numéricos (0 y 1), ya que este, como la mayoría de los modelos, no trabajan con variables de clase categóricas:

In [10]:
# Transformamos True y False en 1 y 0. Para indicar si es spam o no

df['is_spam'] = df['is_spam'].apply(lambda x: 1 if x == True else 0).astype(int)
df.head()

url  is_spam
0  https://briefingday.us8.list-manage.com/unsubs...        1
1                             https://www.hvper.com/        1
2                 https://briefingday.com/m/v4n3i4f3        1
3   https://briefingday.com/n/20200618/m#commentform        0
4                        https://briefingday.com/fan        1

### 2.2 Eliminación de valores repetidos

Podemos contar fácilmente cuántos casos de cada clase tenemos para analizar si el conjunto de datos está equilibrado o no:

In [11]:
row, col = df.shape

print(f"{row} datos y {col} categorías")

print(f"Spam: {len(df.loc[df['is_spam'] == 1])}")

print(f"No Spam: {len(df.loc[df['is_spam'] == 0])}")



2999 datos y 2 categorías
Spam: 696
No Spam: 2303


>Como podemos ver no está equilibrado porque tenemos 696 mensajes de spam y 2303 mensajes de no spam.

Debemos eliminar también los duplicados, si los hubiera:

In [12]:
# Obtain the number of duplicate values in our dataset
def procesado_duplicados(dataframe):

    duplicados = dataframe.duplicated().sum()
    print(f"The number of duplicate values in the dataset is {duplicados}")

    if duplicados > 0:

        print("Después del procesado tenemos...")
        
        dataframe = dataframe.drop_duplicates()

        dataframe = dataframe.reset_index(inplace = False, drop = True)

        final_duplicados = dataframe.duplicated().sum()

        print(f"The number of duplicate values in the dataset is {final_duplicados}")

    return dataframe 

df = procesado_duplicados(df)

The number of duplicate values in the dataset is 630
Después del procesado tenemos...
The number of duplicate values in the dataset is 0


En este caso vemos que se han eliminado más de 600 registros repetidos.


---


### 2.3 Procesamiento del texto

Para poder entrenar el modelo es necesario aplicar antes un proceso de transformación al texto. Comenzamos transformando el texto a minúsculas y eliminando signos de puntuación y caracteres especiales:

In [13]:
# Library which can be used to work with Regular Expressions
# https://www.w3schools.com/python/python_regex.asp

import regex as re

def preprocess_text(text):

    # Eliminar cualquier carácter que no sea una letra (a-z) o un espacio en blanco ( )

    text = re.sub(r'[^a-z ]',#  La construcción [^...] indica negación en las expresiones regulares,
                             # lo que significa "coincide con cualquier cosa que no esté en este conjunto
                   " ", # " ": Esta es la cadena de reemplazo. Cualquier carácter que coincida con el patrón será reemplazado por un espacio.
                     text) 


    # Eliminar espacios en blanco
    text = re.sub(r'\s+[a-zA-Z]\s+', " ", text)
    text = re.sub(r'\^[a-zA-Z]\s+', " ", text)

    # Reducir espacios en blanco múltiples a uno único
    text = re.sub(r'\s+', " ", text.lower())

    # Eliminar tags
    text = re.sub("&lt;/?.*?&gt;"," &lt;&gt; ", text)

    return text.split()

df["url"] = df["url"].apply(preprocess_text)
df.head()

url  is_spam
0  [https, briefingday, us, list, manage, com, un...        1
1                           [https, www, hvper, com]        1
2                    [https, briefingday, com, v, i]        1
3          [https, briefingday, com, m, commentform]        0
4                     [https, briefingday, com, fan]        1

El siguiente paso es la lematización del texto, que es el proceso de simplificación de las palabras a su forma base o canónica, de manera que palabras con diferentes formas, pero el mismo núcleo semántico, se traten como una sola palabra. Por ejemplo, los verbos "corriendo", "corrió" y "corre" serán lematizados a "correr", así como las palabras "mejores" y "mejor" podrían ser lematizadas a "bueno".

Además, aprovechando la lematización, eliminaremos también las stopwords, que son palabras que consideramos irrelevantes para el análisis de texto porque aparece con mucha frecuencia en el lenguaje y no aporta información significativa. Existen dos formas: crear nosotros nuestra propia lista de palabras a eliminar o utilizar librerías externas.

>NOTE: Ambas tareas las llevaremos a cabo con la librería de Python `NLTK`, que es una de las más importantes en términos del NLP:

In [14]:
from nltk import download
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

# Importar WordNet para Lematización
# ================================================================

download("wordnet") # Descarga el conjunto de datos WordNet necesario para realizar la lematización

lemmatizer = WordNetLemmatizer() # Crea una instancia del lematizador WordNet para simplificar palabras a sus formas base.


# Importar y configurar palabras vacías (stopwords)
# ================================================================

download("stopwords") # Descarga la lista de palabras vacías en inglés, como, "the", "is", "and"...

stop_words = stopwords.words("english") # Obtén la lista de palabras vacías en inglés para filtrar palabras irrelevantes.


# Definir una función para lematizar texto
# ================================================================
def lemmatize_text(words, lemmatizer = lemmatizer):
    """
    Lematiza una lista de palabras y filtra palabras irrelevantes.

    Parámetros:
    - words: lista de palabras a procesar.
    - lemmatizer: instancia del lematizador de WordNet (por defecto usa la creada previamente).

    Retorno:
    - Lista de palabras lematizadas, sin palabras vacías y con longitud mayor a 3 caracteres.
    """
    tokens = [lemmatizer.lemmatize(word) for word in words]
    # Lematiza cada pañabra en la lista para obtener su forma base (por ejemplo, "running" -> "run")

    tokens = [word for word in tokens if word not in stop_words]
    # Filtra las palabras vacías para eliminar términos comunes que no aportan significado.
    
    tokens = [word for word in tokens if len(word) > 3]
    # Filtra palabras cuya longitud sea menor o igual a 3 carácteres, ya que suelen ser irrelevantes.


    return tokens # Devuelve la lista procesada de palabras



[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\Fermin\AppData\Roaming\nltk_data...
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Fermin\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping corpora\stopwords.zip.


In [16]:
df.url = df.url.apply(lemmatize_text)
df.head(5)

url  is_spam
0  [http, briefingday, list, manage, unsubscribe]        1
1                                   [http, hvper]        1
2                             [http, briefingday]        1
3                [http, briefingday, commentform]        0
4                             [http, briefingday]        1

Algo muy común una vez tenemos los tokens es representarlos en una nube de palabras. Una nube de palabras es una representación visual de las palabras que componen un texto, donde el tamaño de cada palabra indica su frecuencia o importancia en dicho texto.

Esta representación visual permite identificar rápidamente los términos o conceptos más relevantes o repetidos en un conjunto de datos, ya que las palabras más frecuentes o significativas sobresalen por su tamaño. Podemos implementarla fácilmente utilizando la librería `wordcloud` para Python:

In [19]:
from wordcloud import WordCloud

ModuleNotFoundError: No module named 'wordcloud'